### PR TITLE
New - how to track dependencies

### PR DESCRIPTION
**Pull request by boffbowsh**
This only covers tools for Ruby codebases. Teams using other languages
should contribute their own findings later.

### CODE REVIEW

**Comment by philandstuff** [comment]
This is interesting.  openregister/openregister-java uses the OWASP dependency checker gradle plugin with success - and just checks false positives into the code base.

When I was working on that project, it was up to the PR author to ensure that required false positive configuration was checked in to the same branch.

**Comment by rboulton** [comment]
@heathd Review please wrt usage on Pay.

**Comment by richardTowers** [comment]
We (Verify) are using both greenkeeper and snyk for https://github.com/alphagov/passport-verify#passport-verify (node JS) and finding both to be good.

We've also looked at snyk for a java (gradle) app but it was false-positivesville and it wasn't clear how to mark them, so we gave up.

**Comment by boffbowsh** [comment]
Any thoughts on this, @heathd ?

**Comment by heathd** [comment]
We're using snyk for both maven and nodejs dependencies.

There are some false positive scenarios, but we're gradually getting used to detecting and triaging them.

Overall I feel that snyk is pretty reasonable. 

No tool can replace the need to actually understand and be able to interpret security vulnerabilities. That takes time and effort to build up the knowledge.

I've written some further notes here:

https://docs.google.com/document/d/1J5NK9Dcz0mRKGZ5vBZT5e956G7AToxCd0c0GHMUIKag/edit#

**Comment by boffbowsh** [comment]
The key difference I see here is that Pay allow the creation of `.snyk.yml` whereas GOV.UK don't. I personally disagree with allowing a commercial product to add files to our repositories in order for them to function - I'm surprised they don't store this configuration in their database. I'd welcome other comments on this.

I suggest we agree on allowing or disallowing this, amending the PR if appropriate, then other PRs being raised re: the operating system and Docker vulnerability checkers mentioned by @heathd

**Comment by heathd** [comment]
We mainly have used the .snyk.yml file to suppress warnings about low priority vulnerabilities for which a fix is not yet available. The approach we have been taking is to silence them for a period (a month or so). When the period expires we re-review whether the upstream has made a fix available.

I agree the patching approach needs more careful consideration.

**Comment by soupdragon99** [comment]
@heathd @boffbowsh @richardTowers this PR has been open for a while, are we in a position to publish this as guidance for the GDS Way, after a tech writer review?

**Comment by boffbowsh** [comment]
I believe @heathd had a conversation recently within Pay about dependencies that is probably pertinent to this

**Comment by boffbowsh** [comment]
I believe the sentiment from @heathd after extended use of Snyk was that it wasn't working for them. I think this can be merged as-is and further detail added later @soupdragon99

**Comment by soupdragon99** [comment]
@boffbowsh thanks for updating, I'll need review this before we can publish.

**Comment by boffbowsh** [comment]
The security audit gem mentioned only works for Ruby, so it doesn't match the requirements of the guidance. In which case, I don't think this guidance has any further value. The main value was in us documenting why we don't use Snyk.

**Comment by soupdragon99** [comment]
Thanks @boffbowsh it'd be good to catch up on this to help my understanding of what we're saying here. In the meantime, I'll reinstate the content around Synk as it's a key component.

**Comment by boffbowsh** [comment]
Do you mean 2i from me or another Technical Writer? We had it confirmed from Rosalie that your review is all that's needed from a tech writer POV.

File: source/standards/tracking-dependencies.html.md.erb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+You should use the [GOV.UK Gem Security Checker][] alongside your regular code checks. This will help your team move any code vulnerabilities into the team’s work backlog. You can then:
+
+* prioritise fixes relative to other project work
+* address vulnerabilities in private before making the fix public
+
+## Dependency management tools
+
+A common way to manage third party dependencies is to use a dependency management tool. [The Service Manual][] has some suggestions about [how to work with third party code][] and about some common dependency management tools.
+
+## Case study - Snyk
```
[the remark this answers is not in the document]
Comment (boffbowsh) [reply source/standards/tracking-dependencies.html.md.erb]: We don't, it's free.

[the remark this answers is not in the document]
Comment (soupdragon99) [reply source/standards/tracking-dependencies.html.md.erb]: Thanks

**Pull request merged**